File: spice_sim/scripts/iqmethod.m
function [mag, phaserad] = iqmethod(sig, len, f0, fs, pperiod) %#codegen
% KEEP THIS IN MIND
% This calculates the phasor from a reference signal of unknown phase
% relative to the reference signal: this means that so long as we compute
% the phasor of the input AND output, we can obtain the phasor of the
% output relative to the input, which is effectively what we want.
    %% ---- GET IN PHASE AND QUADRATURE REF SIGNALS ---- %%
    t = 1 : len;
    ang_step = (2.0 * pi * t * f0) / ( fs) ;
    inphase_sig = sin(ang_step);
    quadrature_sig = cos(ang_step);

    %% ---- SUM UP ---- %%
    size(sig)
    size(inphase_sig)
    I_sig = inphase_sig .* sig;
    Q_sig = quadrature_sig .* sig;

    I = sum(I_sig) / len;
    Q = sum(Q_sig) / len;

    mag = sqrt(I^2 + Q^2);
    phaserad = atan2(Q, I);

end

File: InstrumentUI/iqmethod.py
import numpy as np


def phasor_1sig(sig, f0, fs):
    I = Q = 0
    for i in range(len(sig)):
        I += sig[i] * np.cos(2*np.pi*i *f0/fs )
        Q += sig[i] * np.sin(2*np.pi*i *f0/fs)

    I /= len(sig)
    Q /= len(sig)

    mag = np.sqrt(I*I + Q*Q)
    phaserad = np.atan2(Q, I)

    return mag * (np.cos(phaserad) + 1j*np.sin(phaserad))

def phasor_2sig(sig, ref, f0, fs):
    return phasor_1sig(sig, f0, fs) / phasor_1sig(ref, f0, fs)

def phasors2Zx(v1, v2, Rref):
    return Rref * v2 / (v1 - v2)